crates/telio-traversal/src/endpoint_state.rs

```diff
@@ -0,0 +1,120 @@
+use std::fmt;
+use telio_utils::telio_log_warn;
+
+#[derive(Debug, Default, PartialEq)]
+pub enum EndpointState {
+    #[default]
+    Disconnected,
+    EndpointGathering,
+    Ping,
+    Published,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Event {
+    SendCallMeMaybeRequest,
+    ReceiveCallMeMaybeResponse,
+    Publish,
+    Timeout,
+    EndpointGone,
+}
+
+#[derive(Default, Debug)]
+pub struct EndpointStateMachine {
+    state: EndpointState,
+    last_event: Option<Event>,
+}
+
+impl EndpointStateMachine {
+    pub fn handle_event(&mut self, event: Event) {
+        match (&self.state, &event) {
+            (EndpointState::Disconnected, Event::SendCallMeMaybeRequest) => {
+                self.last_event = Some(event);
+                self.state = EndpointState::EndpointGathering;
+            }
+
+            (EndpointState::EndpointGathering, Event::ReceiveCallMeMaybeResponse) => {
+                self.last_event = Some(event);
+                self.state = EndpointState::Ping;
+            }
+
+            (EndpointState::Ping, Event::Publish) => {
+                self.last_event = Some(event);
+                self.state = EndpointState::Published;
+            }
+
+            (EndpointState::EndpointGathering, Event::Timeout)
+            | (EndpointState::Ping, Event::Timeout) => {
+                self.last_event = Some(event);
+                self.state = EndpointState::Disconnected;
+            }
+
+            (EndpointState::Published, Event::EndpointGone) => {
+                self.last_event = Some(event);
+                self.state = EndpointState::Disconnected;
+            }
+
+            (_, event) => {
+                telio_log_warn!("Invalid state transition {:?} -> {:?}", &self.state, event);
+            }
+        }
+    }
+
+    pub fn get(&self) -> &EndpointState {
+        &self.state
+    }
+
+    pub fn last_event(&self) -> &Option<Event> {
+        &self.last_event
+    }
+}
+
+impl PartialEq<EndpointState> for EndpointStateMachine {
+    fn eq(&self, other: &EndpointState) -> bool {
+        &self.state == other
+    }
+}
+
+impl PartialEq<(EndpointState, Option<Event>)> for EndpointStateMachine {
+    fn eq(&self, other: &(EndpointState, Option<Event>)) -> bool {
+        self.state == other.0 && self.last_event == other.1
+    }
+}
+
+impl fmt::Display for EndpointStateMachine {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self.state)
+    }
+}
```

Comment:
Was this automatically implemented by the `sm` macro? And if so, does this differ from the old implementation? I could very well be mistaken because I haven't looked into it at all, but I would imagine that if this existed before, it would include the event that led to the current state just by nature of how `sm` creates the enum(s)

Reply:
I'm not sure how `sm` handled this but I've implemented this trait so that the logs end up unchanged with what was before:
`2024-09-03T13:28:54.881583Z  INFO telio_traversal::connectivity_check::cross_ping_check: 800: Node's "JRAb...n0s=" EP 200.225.115.103:24020 state transition Published -> EndpointGone`

Reply:
Okay, that's all good then